runtime/caml/mlvalues.h

```diff
@@ -98,47 +99,50 @@ bits  31    10 9     8 7   0
 
 For 64-bit architectures:
 
-     +--------+-------+-----+
-     | wosize | color | tag |
-     +--------+-------+-----+
-bits  63    10 9     8 7   0
+     +----------+--------+-------+-----+
+     | reserved | wosize | color | tag |
+     +----------+--------+-------+-----+
+bits  63    64-R 63-R  10 9     8 7   0
 
-For x86-64 with Spacetime profiling:
-  P = PROFINFO_WIDTH (as set by "configure", currently 26 bits, giving a
-    maximum block size of just under 4Gb)
-     +----------------+----------------+-------------+
-     | profiling info | wosize         | color | tag |
-     +----------------+----------------+-------------+
-bits  63        (64-P) (63-P)        10 9     8 7   0
+where 0 <= R <= 31 is HEADER_RESERVED_BITS, set with the
+--enable-reserved-header-bits=R argument to configure.
 
 */
 
-#define Tag_hd(hd) ((tag_t) ((hd) & 0xFF))
-
-#define Gen_profinfo_shift(width) (64 - (width))
-#define Gen_profinfo_mask(width) ((1ull << (width)) - 1ull)
-#define Gen_profinfo_hd(width, hd) \
-  (((mlsize_t) ((hd) >> (Gen_profinfo_shift(width)))) \
-   & (Gen_profinfo_mask(width)))
-
-#ifdef WITH_PROFINFO
-#define PROFINFO_SHIFT (Gen_profinfo_shift(PROFINFO_WIDTH))
-#define PROFINFO_MASK (Gen_profinfo_mask(PROFINFO_WIDTH))
-/* Use NO_PROFINFO to debug problems with profinfo macros */
-#define NO_PROFINFO 0xff
-#define Hd_no_profinfo(hd) ((hd) & ~(PROFINFO_MASK << PROFINFO_SHIFT))
-#define Wosize_hd(hd) ((mlsize_t) ((Hd_no_profinfo(hd)) >> 10))
-#define Profinfo_hd(hd) (Gen_profinfo_hd(PROFINFO_WIDTH, hd))
-#else
-#define NO_PROFINFO 0
-#define Wosize_hd(hd) ((mlsize_t) ((hd) >> 10))
-#define Profinfo_hd(hd) NO_PROFINFO
-#endif /* WITH_PROFINFO */
+#define HEADER_BITS (sizeof(header_t) * CHAR_BIT)
+
+#define HEADER_TAG_BITS 8
+#define HEADER_TAG_MASK ((1ull << HEADER_TAG_BITS) - 1ull)
+
+#define HEADER_COLOR_BITS 2
+#define HEADER_COLOR_SHIFT HEADER_TAG_BITS
+#define HEADER_COLOR_MASK ((1ull << HEADER_COLOR_BITS) - 1ull)
+
+#define HEADER_WOSIZE_BITS (HEADER_BITS - HEADER_TAG_BITS - HEADER_COLOR_BITS - HEADER_RESERVED_BITS)
+#define HEADER_WOSIZE_SHIFT (HEADER_COLOR_SHIFT  + HEADER_COLOR_BITS)
+#define HEADER_WOSIZE_MASK ((1ull << HEADER_WOSIZE_BITS) - 1ull)
+
+#define Tag_hd(hd) ((tag_t) ((hd) & HEADER_TAG_MASK))
+#define Wosize_hd(hd) ((mlsize_t) (((hd) >> HEADER_WOSIZE_SHIFT) & HEADER_WOSIZE_MASK))
+
+/* These macros are ugly because the compiler complains about (0 ? (x
+ * << 64) : 0), so I have to conditionalize more. */
```

Comment:
Have you considered using

```c
#if HEADER_RESERVED_BITS == 0
#define Reserved_hd(hd) ...
#define Hd_reserved(res) ...
#else
#define Reserved_hd(hd) ...
#define Hd_reserved(res) ...
#endif
```
?

(One downside of the current approach is that you do expose the tricky-to-use _SHIFT variable to the rest of the codebase.)

Reply:
Yes, I thought of it, but generally try to avoid `#if` where I can (somewhere I have a massive rant I wrote in about 1994 entitled "#if Considered Harmful"). I guess this isn't too bad though.

Reply:
I see that you finally pushed a commit along these lines. For the record, I'm happy to let you stick to your preference if you have a strong one, changing this was not a hard requirement in any way. I do find the new version easier to read, and I like that the potentially-unsafe constant is gone from the user scope, but you obviously have more experience than I do on C-level hackery and I'm happy to follow your intuition on what is best.